### PR TITLE
feat: add web browser native module

### PR DESCRIPTION
**Pull request by cshfang**
#### Description of changes

This PR adds a react native module which allows auth sessions to be opened via:
* ASWebAuthenticationSession on iOS
* Chrome tabs on Android

#### Description of how you validated changes
Tested with a v6 POC

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/rtn-web-browser/webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+module.exports = {
```
Comment (HuiSF) [inline packages/rtn-web-browser/webpack.config.js:1]: Considering this is a JS + RN Bridge + Native module package. Do we still need to generate a webpack bundle for the js portion?

File: packages/rtn-web-browser/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { NativeModules } from 'react-native';
+import { WebBrowserNativeModule } from './types';
+import { openAuthSessionAsync } from './apis/openAuthSessionAsync';
+
+const module: WebBrowserNativeModule = NativeModules.AmplifyRTNWebBrowser;
```
Comment (HuiSF) [inline packages/rtn-web-browser/src/index.ts:8]: Maybe can reused the `module` at line 7 in `openAuthSessionAsync.ts`.

Also do we need to handle and throw when `NativeModules.AmplifyRTNWebBrowser` is `undefined`? (for any reason that autolinking failed)

File: packages/rtn-web-browser/src/apis/openAuthSessionAsync.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+				// wait for app to return some other way, resulting in null
+				getAppStatePromise(),
+			]),
+			// open chrome tab
+			module.openAuthSessionAsync(url),
+		]);
+		return redirectUrl;
+	} finally {
+		appStateListener.remove();
+		redirectListener.remove();
```
Comment (HuiSF) [inline packages/rtn-web-browser/src/apis/openAuthSessionAsync.ts:45]: The type here would be `<ListenerHandlerType> | undefined`. Should be (if ts restrict gets enabled)

```suggestion
		appStateListener?.remove();
		redirectListener?.remove();
```

File: packages/rtn-web-browser/android/src/main/kotlin/com/amazonaws/amplify/rtnwebbrowser/WebBrowserModule.kt [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                val customTabsIntent = CustomTabsIntent.Builder(connection?.getSession()).build()
+                customTabsIntent.intent.setPackage(it).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+                customTabsIntent.launchUrl(reactApplicationContext, Uri.parse(uriStr))
+            } ?: run {
+                promise.reject(Throwable("No eligible browser found on device"))
+            }
+        } catch (e: Exception) {
+            promise.reject(e)
+        }
+		promise.resolve(null)
```
Comment (HuiSF) [inline packages/rtn-web-browser/android/src/main/kotlin/com/amazonaws/amplify/rtnwebbrowser/WebBrowserModule.kt:53]: nit: indentation seems off

File: packages/rtn-web-browser/CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
+# Change Log
+
+All notable changes to this project will be documented in this file.
+See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.
```
Comment (HuiSF) [inline packages/rtn-web-browser/CHANGELOG.md:4]: nit:

```suggestion
## 1.1.0

* Initial implementation
```

File: packages/rtn-web-browser/ios/AmplifyRTNWebBrowser.swift [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    }
+}
+
+@objc(AmplifyRTNWebBrowser)
+class AmplifyRTNWebBrowser: NSObject {
+    var webBrowserAuthSession: ASWebAuthenticationSession?
+    private let presentationContextProvider = PresentationContextProvider()
+    
+    private func isUrlValid(url: URL) -> Bool {
+        return url.scheme == "http" || url.scheme == "https"
```
Comment (5d) [inline packages/rtn-web-browser/ios/AmplifyRTNWebBrowser.swift:19]: nit: maybe using caseInsensitiveCompare
```suggestion
     url.scheme != nil && 
     ["http", "https"].contains { $0.caseInsensitiveCompare(url.scheme!) == .orderedSame } 
```

**Review by 5d — APPROVED** [review]

File: packages/rtn-web-browser/android/src/main/kotlin/com/amazonaws/amplify/rtnwebbrowser/WebBrowserModule.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                val customTabsIntent = CustomTabsIntent.Builder(connection?.getSession()).build()
+                customTabsIntent.intent.setPackage(it).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+                customTabsIntent.launchUrl(reactApplicationContext, Uri.parse(uriStr))
+            } ?: run {
+                promise.reject(Throwable("No eligible browser found on device"))
+            }
+        } catch (e: Exception) {
+            promise.reject(e)
+        }
+				promise.resolve(null)
```
Comment (jimblanc) [inline packages/rtn-web-browser/android/src/main/kotlin/com/amazonaws/amplify/rtnwebbrowser/WebBrowserModule.kt:55]: Nit: tabbing error

**Review by jimblanc — APPROVED** [review]
lgtm

**Review by jimblanc — APPROVED** [review]

**Review by HuiSF — APPROVED** [review]

**Pull request merged**